lib/rgit/cli.rb

```diff
@@ -55,13 +258,275 @@ def format_options(option_hash)
         enum: ["false", "true", "umask", "group", "all", "world", "everybody"],
         desc: @init_descriptions[:shared]
       }
-    def init
+    def init(directory = Pathname.pwd)
+      # Get directory as pathname, default being current location
+      directory = Pathname.new(directory)
+      unless directory.exist? # make the directory if it doesn't exist
+        directory.mkpath
+      end
 
+      # Find parent subrepo/repo if it exists
+      git_parent, subrepo_parent = lowest_above(directory)
+      # If we have a parent, and that parent is not us
+      # -> initialize a subrepo
+      # else -> initialize a repo
+      # dir != parent check is needed since user may "re-init" a repo
+      if (git_parent) && directory != git_parent
+        init_subrepo(git_parent, subrepo_parent, directory)
+      else
+        init_repo(directory, options)
+      end
     end
 
     desc "test", "A test function"
     def test(*args)
-      puts args
+      j = args.class
+    end
+  end
+end
+
+module Impl
+  VERSION = Rgit::VERSION
+  include Wrappers
+  require "pty"
+  def format_options(option_hash)
+    result = ""
+    option_hash.each { |k, v|
+      key_str = " --#{k.to_s.gsub("_", "-")}"
+      case v # v is truthy in all cases except: nil, false
+        when [true, false].include?(v) # it's a boolean
+          result << key_str
+        when String
+          result << key_str << "=#{v}"
+        when Fixnum
+          result << key_str << "=#{v}"
+        when Hash
+          v.each { |key, val|
+            result << key_str << " #{key}=#{val}"
+          }
+      end
+    }
+    result
+  end
+
+  def lowest_repo_above(start_dir)
+    res_dir = nil
+    start_dir.realpath.ascend { |dir|
+      if dir.has_child? ".git"
+        res_dir = dir
+        break
+      end
+    }
+    res_dir
+  end
+
+  def lowest_gitrepo_above(start_dir)
+    res_dir = nil
+    start_dir.realpath.ascend{ |dir|
+      if dir.has_child? ".git"
+        res_dir = dir
+        break
+      end
+    }
+    res_dir
+  end
+
+  def lowest_above(start_dir)
+    lowest_gitrepo = lowest_gitrepo_above(start_dir)
+    lowest_subrepo = nil
+    if lowest_gitrepo
+      branches = Branches.new(lowest_gitrepo)
+      lowest_subrepo = branches.path_for_branch(branches.branch_for_path(start_dir))
+    end
+    [lowest_gitrepo, lowest_subrepo]
+  end
+
+  # Calling git & shells
+  def git_command(cmd, opt_str, dir)
+    run_in_shell("git #{cmd} #{opt_str}", dir)
+  end
+
+  def run_in_shell(cmd, dir)
+    output = `cd #{dir}; #{cmd}; cd; cd - ;`
+    output
+  end
+
+  # Opens a virtual shell at the specified
+  # dir and runs the given cmd
+  def run_in_pty(cmd, dir)
+    result = ''
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.